src/sftp.rs

```diff
@@ -349,18 +359,30 @@ impl Sftp {
     pub fn unlink(&self, file: &Path) -> Result<(), Error> {
         let file = util::path2bytes(file)?;
         self.rc(unsafe {
-            raw::libssh2_sftp_unlink_ex(self.raw, file.as_ptr() as *const _, file.len() as c_uint)
+            raw::libssh2_sftp_unlink_ex(
+                self.unwrap_raw_or_err()?,
+                file.as_ptr() as *const _,
+                file.len() as c_uint,
+            )
         })
     }
 
     /// Peel off the last error to happen on this SFTP instance.
     pub fn last_error(&self) -> Error {
-        let code = unsafe { raw::libssh2_sftp_last_error(self.raw) };
+        let raw = match self.unwrap_raw_or_err() {
+            Ok(raw) => raw,
+            Err(e) => return e,
+        };
+        let code = unsafe { raw::libssh2_sftp_last_error(raw) };
         Error::from_errno(code as c_int)
     }
 
-    fn last_session_error(&self) -> Option<Error> {
-        Error::last_error_raw(self._sess.raw)
+    fn last_session_error(&self) -> Result<Option<Error>, Error> {
```

Comment:
This is a tricky function signature!
I think the only use of this also uses `unwrap_or_else(Error::unknown)`, so how about we make this a little simpler?

```rust
fn last_session_error(&self) -> Error {
   if let Some(inner) = self.inner.as_ref() {
      Error::last_error_raw(inner.sess.raw).unwrap_or_else(Error::unknown)
   } else {
      Error::from_errno(raw::LIBSSH2_ERROR_BAD_USE)
   }
}
```